closetCleaner lib

Fox214 opened the pull request:
library addon to report gear usage across specified job files.

Commit:
closetCleaner

gearswap library to report gear usage across files and indicate what in your inventory is unused.

Commit:
closetCleaner lib

library addon to report gear usage across specified job files.

Fox214 commented:
any idea when this might get looked at?

z16 commented on addons/GearSwap/libs/ccConfig.lua:
```diff
@@ -0,0 +1,30 @@
+-- Jobs you want to execute with, recomment put all active jobs you have lua for will look for <job>.lua or <playername>_<job>.lua files
```
I don't know how GS usually does this, but normally user files would go into the `GearSwap/data/` folder. That folder is special in that it won't be updated and overwritten. For example, if you ever wanted to update this file with new default values, the file would be overwritten with the new default file for all users who modified it.

Two options you can do then, either move this file to the aforementioned `data` folder, then it will be downloaded initially but ignored in any future updates (can make a subfolder for cc there as well), or forgo this file entirely. I know you want to show people how to use cc through this file, but I think that would work better if you added a blank file to the `data` folder, and keep this file here as an example, and name it as such (possibly refer to it in a readme or w/e). That way people trying it for the first time without looking at the config file will not experience weird behavior, which they will complain to you (and us) about in the end.

z16 commented on addons/GearSwap/libs/ccConfig.lua:
```diff
@@ -0,0 +1,30 @@
+-- Put any items in your inventory here you don't want to show up in the final report
+-- recommended for furniture, food, meds, pop items or any gear you know you want to keep for some reason
+-- supports lua regex syntax, put each separate one as its own entry in the table
+-- see: https://www.lua.org/pil/20.2.html
```
I would suggest using wildcard matching instead of Lua's string pattern matching. You would need to use the `windower.wc_match` function instead of `string.match` below, but people are more familiar with it and it's generally simpler to use (you actually used it by accident on `"Deed of*"` below, which is not a valid Lua string pattern (or well, it is, but probably not the one you intended it to be)). The wildcard matching is in many cases also more powerful, since it allows alternation with `|`, which Lua string patterns do not.

z16 commented on addons/GearSwap/libs/ccConfig.lua:
```diff
@@ -0,0 +1,30 @@
+-- supports lua regex syntax, put each separate one as its own entry in the table
+-- see: https://www.lua.org/pil/20.2.html
+ccignore = { "Rem's Tale*", "Storage Slip %d+", "Deed of*", "%a+ Virtue", "Dragua's Scale",
+			"Glittering Yarn", "Dim. Ring*", "Cupboard", ".*VCS.*", ".*Abjuration.*", "%a+ Organ",
```
Use spaces instead of tabs please. Superficially it's not too important, but we've been treating it as a policy. You can configure most editors to use spaces when you press the tab key.

The same goes for the other file.

z16 commented on addons/GearSwap/libs/closetCleaner.lua:
```diff
@@ -0,0 +1,379 @@
+--Copyright (c) 2016~2017, Brimstone
```
```
--Copyright © 2016-2017, Brimstone
```

z16 commented on addons/GearSwap/libs/closetCleaner.lua:
```diff
@@ -0,0 +1,379 @@
+--    * Redistributions in binary form must reproduce the above copyright
+--      notice, this list of conditions and the following disclaimer in the
+--      documentation and/or other materials provided with the distribution.
+--    * Neither the name of <addon name> nor the
```
Use `closetCleaner` or whatever you want this library's "official" name to be.

z16 commented on addons/GearSwap/libs/closetCleaner.lua:
```diff
@@ -0,0 +1,379 @@
+--THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+--ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+--WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+--DISCLAIMED. IN NO EVENT SHALL <your name> BE LIABLE FOR ANY
```
Use `Brimstone` here.

z16 commented on addons/GearSwap/libs/closetCleaner.lua:
```diff
@@ -0,0 +1,379 @@
+register_unhandled_command(function(...)
+    local cmds = {...}
+    for _,v in ipairs(cmds) do
+        if S{'closetCleaner','cc'}:contains(v:lower()) then
```
If you do `v:lower()` it will never match `closetCleaner`. Should lowercase that as well.

z16 commented on addons/GearSwap/libs/closetCleaner.lua:
```diff
@@ -0,0 +1,379 @@
+			cc.sandbox.gsGear = T{}
+			for k,v in pairs(gearswap.res.items) do
+				cc.sandbox.itemsBylongName[gearswap.res.items[k].enl:lower()] = k
+				cc.sandbox.itemsByName[gearswap.res.items[k].en:lower()] = k
```
A note on using our resources. Please use `english` and `english_log` instead of `en` and `enl`, despite what it shows in the file. We use the short version in the file primarily to save on file size since the resources can get huge in some cases.

Also, I don't know how well that works with other GearSwap features, but consider using `name` and `name_log` instead of either. That will make it so that Japanese players can write their item names in Japanese and it will still match them, since `name` expands to `english` for English clients, and `japanese` for Japanese ones.

z16 commented on addons/GearSwap/libs/closetCleaner.lua:
```diff
@@ -0,0 +1,379 @@
+end
+
+-- recursively copies a table into a new table, this avoids dealing with references so proper tallies can be done
+function cc.deepcopy(orig)
```
Fwiw there's already a deepcopy function in our tables library, which works on `T{}`-tables by just writing `orig:copy(true)` (`true` indicated a deep instead of shallow copy). For non-`T{}`-tables you can do `table.copy(orig, true)`.

z16 replied:
Actually I just realized the copy is deep by default (i.e. you have to explicitly specify `false` to make it shallow). So just writing `orig:copy()` should do (or `table.copy(orig)` for non-`T{}`-tables).

z16 commented on addons/GearSwap/libs/closetCleaner.lua:
```diff
@@ -0,0 +1,379 @@
+        table.sort(keys, function(a,b) return order(t, a, b) end)
+    else
+        table.sort(keys)
+    end
```
We do not have a sorted iterator, but the above few lines could be replaced by our library functions:
```lua
local keys = t:keyset():sort()
```

Or, if `t` is not a `T{}`-table:
```lua
local keys = table.sort(table.keyset(t))
```

z16 reviewed (changes requested):
I added a review, but keep in mind that I haven't been playing since GS was a thing and never used it or any libraries for it, so my knowledge on it is very limited and I didn't actually test your code. I basically only commented on Lua-specific things I noticed or things regarding the Windower API or overall design. I'll trust that you know and use GS well enough yourself to know if this works or doesn't :)

Most of the things I said were just suggestions, the only changes we need to have are the changes to the copyright notice I mentioned and the tabs thing. I'll leave the rest up to your discretion.

Commit:
Merge branch 'dev' of https://github.com/Fox214/Lua into dev

Added copy write info, a few other minor changes, moved a blank ccConfig.lua to data directory and added example file.

Fox214 commented:
Apologies for taking a month to get back to this, I've taken your first few recommendations, I was not able to get the windower deep copy or table sorts to work so I left it alone.

I thought I had np++ configed for space tabs, but Apparently not, I've fixed it, but will only slowly change the code. Guess that's what happens when I code on 3-4 computers. 

Also I've noticed a bug that the Mote-Treasurehunter seems to try to do stuff after the files are executed. (its at the bottom of the file windoer.raw_register_event(action, on_action_for_th) where there is uninitialized values, just doing //lua r gearswap fixes it, but it echos errata to the console otherwise. Not sure how to fix it.

trv6 commented:
You can replace all tabs with spaces in notepad++ pretty quickly.
Press `<ctrl h>` to open the replace menu. Select the extended search mode button, then replace all instances of `\t` with `    ` (four spaces).

z16 commented:
Unfortunately I'm not proficient enough with GS to answer that. A bit curious why our `copy`/`sort` isn't working for you, would be cool if you could give an example,

And feel free to commit the new changes you made and push them to your remote branch, they will automatically appear in this pull request, then we can keep reviewing it.

Commit:
bug fixes

Converted tabs to spaces, fixed a bug where sets were not getting initialized properly.

Fox214 commented:
I got the deep copy to work, I didn't bother with the other for now at 
least because it looked more complex and my brain hurts.

I fixed the tabs (can't believe I didn't think to use a regex).

Also fixed a functional bug in the code, and recommited to dev let me 
know if you can't see it.

------ Original Message ------
From: "Stjepan Bakrac" <notifications@github.com>
To: "Windower/Lua" <Lua@noreply.github.com>
Cc: "Fox214" <steven.fox@zoho.com>; "Author" <author@noreply.github.com>
Sent: 10/10/2017 4:59:09 AM
Subject: Re: [Windower/Lua] closetCleaner lib (#1558)

>Unfortunately I'm not proficient enough with GS to answer that. A bit 
>curious why our copy/sort isn't working for you, would be cool if you 
>could give an example,
>
>And feel free to commit the new changes you made and push them to your 
>remote branch, they will automatically appear in this pull request, 
>then we can keep reviewing it.
>
>—
>You are receiving this because you authored the thread.
>Reply to this email directly, view it on GitHub 
><https://github.com/Windower/Lua/pull/1558#issuecomment-335407571>, or 
>mute the thread 
><https://github.com/notifications/unsubscribe-auth/ARqYD-AwrAl6gT0QlktCm-Q0cd8ce30cks5sqzHdgaJpZM4O8mdc>.
>

Fox214 commented:
Does this look ok now?

z16 commented:
Oh sorry, I had the tab open but forgot to look at it... I'll check it out when I wake up.

z16 commented:
I looked through it and it seems largely the same as it was before. Are you sure you committed everything? It needs to be committed and pushed to the same branch from which you opened the pull request (`Fox214:dev`).

Fox214 commented:
I see them here: 
https://github.com/Windower/Lua/pull/1558/commits/27c468ff47437915179b255ab475fe0d6ec9f599

Says this:
          closetCleaner lib #1558
Open
Fox214 <https://github.com/Fox214> wants to merge 3 commits into 
Windower:dev from Fox214:dev

------ Original Message ------
From: "Stjepan Bakrac" <notifications@github.com>
To: "Windower/Lua" <Lua@noreply.github.com>
Cc: "Fox214" <steven.fox@zoho.com>; "Author" <author@noreply.github.com>
Sent: 11/9/2017 8:29:02 AM
Subject: Re: [Windower/Lua] closetCleaner lib (#1558)

>I looked through it and it seems largely the same as it was before. Are 
>you sure you committed everything? It needs to be committed and pushed 
>to the same branch from which you opened the pull request (Fox214:dev).
>
>—
>You are receiving this because you authored the thread.
>Reply to this email directly, view it on GitHub 
><https://github.com/Windower/Lua/pull/1558#issuecomment-343154313>, or 
>mute the thread 
><https://github.com/notifications/unsubscribe-auth/ARqYDwq4sAmuQ5BlpCejxQ-7d5p7qb00ks5s0v4egaJpZM4O8mdc>.
>

cairthenn commented on addons/GearSwap/libs/closetCleaner.lua:
```diff
@@ -265,7 +265,6 @@ function cc.list_sets(t, f)
 							if write_sets[itemid] == nil then
```
tabs :cry:

Fox214 commented:
my bad, i think I referenced the wrong commit, I had a small bug fix and 
all the tabs fixed here:
https://github.com/Windower/Lua/pull/1558/commits/b042ea247e5ac7d50fdcc10704aad78e27a6e149


------ Original Message ------
From: "Mark" <notifications@github.com>
To: "Windower/Lua" <Lua@noreply.github.com>
Cc: "Fox214" <steven.fox@zoho.com>; "Author" <author@noreply.github.com>
Sent: 11/9/2017 10:53:43 PM
Subject: Re: [Windower/Lua] closetCleaner lib (#1558)

>@cairface commented on this pull request.
>
>--------------------------------------------------------------------------------
>In addons/GearSwap/libs/closetCleaner.lua 
><https://github.com/Windower/Lua/pull/1558#discussion_r150149429>:
>
> > @@ -265,7 +265,6 @@ function cc.list_sets(t, f)
>  							if write_sets[itemid] == nil then
>
>tabs 😢
>
>—
>You are receiving this because you authored the thread.
>Reply to this email directly, view it on GitHub 
><https://github.com/Windower/Lua/pull/1558#pullrequestreview-75658520>, 
>or mute the thread 
><https://github.com/notifications/unsubscribe-auth/ARqYD55C3mQi9LTMUUjqhWOqfjiqzxyNks5s08jGgaJpZM4O8mdc>.
>

z16 commented on addons/GearSwap/data/ccConfig.lua:
```diff
@@ -0,0 +1,14 @@
+-- List bags you want to not check against, needs to match "Location" column in <player>_report.txt
+skipBags = S{ 'Temporary' }
+-- this prints out the _sets _ignored and _inventory files
+ccDebug = false
```
This file shouldn't be there. I'd trust people to be competent enough to look at the example config and figure it out, although I would also move that out of the data folder. The problem with the data folder is that anything in there will never be updated. So if you ever adjust or expand the mechanism people's files will not be overwritten (because they're meant to be *user* created/edited files). This includes the example file.

z16 commented on addons/GearSwap/data/ccConfig.lua:
```diff
@@ -0,0 +1,14 @@
+-- Set to nil or delete for unlimited
+ccmaxuse = nil
+-- List bags you want to not check against, needs to match "Location" column in <player>_report.txt
+skipBags = S{ 'Temporary' }
```
All the other values in here have `cc` prefixed (why?), but this one doesn't. Is this an oversight?

z16 commented on addons/GearSwap/data/ccConfig_example.lua:
```diff
@@ -0,0 +1,30 @@
+            "Carver's Torque", "Kabuto.kazari", "Etched Memory", "Trbl. Hatchet", "Maat's Cap", "Trbl. Pickaxe",
+            "Olduum Ring", "Linkpearl", "Caliber Ring", "Vana'clock", "Signal Pearl", "Windy Greens", "Nexus Cape",
+            "Wastebasket", "Sickle", "Carpenter's Gloves", "Shinobi.Tabi", "Guardian Board", "%a+ Pie", "Pet Food %a+",
+            "Reraise %a+", "P. %a+ Card", "Diorite", "Abdhaljs Seal", "Kin's Scale", "*Curry*", "Crepe B. Helene",
```
Personally I would use one matching mechanism, i.e. `windower.wc_match` or `string.match`, not mix both. I'd recommend wildcard matching over Lua patterns when dealing with users.

z16 commented on addons/GearSwap/libs/closetCleaner.lua:
```diff
@@ -0,0 +1,367 @@
+--Copyright (c) 2016-2017, Brimstone
```
Please adjust `(c)` to `©`. Are you `Brimstone`, or is that a copy/paste oversight?

z16 commented on addons/GearSwap/libs/closetCleaner.lua:
```diff
@@ -0,0 +1,367 @@
+-- _addon.name = 'closetCleaner'
+-- _addon.version = '0.9'
+-- _addon.author = 'Brimstone'
+-- _addon.commands = {'cc','closetCleaner'}
```
Was this originally its own addon, and that's why it's commented out? Any reason to keep it around?

z16 commented on addons/GearSwap/libs/closetCleaner.lua:
```diff
@@ -0,0 +1,367 @@
+local cc = {}
+cc.sandbox = {}
+cc.sandbox.windower = setmetatable({}, {__index = windower})
+cc.sandbox.windower.register_event = functions.empty()
```
Are you trying to disable `register_event`? If so, you'd want `functions.empty`, not `functions.empty()`. The latter is the same as using `nil`.

Any particular reason to do that? I would probably throw an error instead, so people actually know not to use it, instead of nothing happening:
```lua
cc.sandbox.windower.register_event = error:apply('Registering events disabled for CC')
```

trv6 replied:
The error might lead to confusion when people run the script for the first time. 

The library only loads the user files to get to the gear sets, but `register_event` (and a few other functions, I think?) would make persistent changes within the environment of the user's current gearswap file. If they weren't blocked, you'd have to reload gearswap after running the script, but there's not much need to know that the library is disabling them within its own sandbox.

z16 commented on addons/GearSwap/libs/closetCleaner.lua:
```diff
@@ -0,0 +1,367 @@
+
+register_unhandled_command(function(...)
+    local cmds = {...}
+    for _,v in ipairs(cmds) do
```
Are you sure this is what you want? This would trigger even if `cc` was a third parameter in an unhandled command that was meant to be handled by something else.

I'd change the function signature and setup to this:
```lua
register_unhandled_command(function(command)
    command = command and command:lower() or nil
    if command ~= 'cc and command ~= 'closetcleaner' then
        return
    end

    --[[ rest of the function ]]
end)
```

z16 commented on addons/GearSwap/libs/closetCleaner.lua:
```diff
@@ -0,0 +1,367 @@
+            end
+            cc.sandbox.jobs = {}
+            for k,v in pairs(gearswap.res.jobs) do
+                cc.sandbox.jobs[gearswap.res.jobs[k].ens] = k
```
Please use `english_short` instead of `en`.

z16 commented on addons/GearSwap/libs/closetCleaner.lua:
```diff
@@ -0,0 +1,367 @@
+            if not windower.dir_exists(windower.addon_path..'report') then
+                windower.create_dir(windower.addon_path..'report')
+            end
+            local path = windower.addon_path:gsub('\\','/')
```
Our of curiosity, did you have issues using `\` over `/`?

z16 commented on addons/GearSwap/libs/closetCleaner.lua:
```diff
@@ -0,0 +1,367 @@
+    cc.sandbox.windower = setmetatable({}, {__index = windower})
+    cc.sandbox.windower.register_event = functions.empty()
+    cc.sandbox.windower.raw_register_event = functions.empty()
+    cc.sandbox.windower.register_unhandled_command = functions.empty()
```
See above about `function.empty`.

z16 commented on addons/GearSwap/libs/closetCleaner.lua:
```diff
@@ -0,0 +1,367 @@
+    cc.sandbox.windower.register_event = functions.empty()
+    cc.sandbox.windower.raw_register_event = functions.empty()
+    cc.sandbox.windower.register_unhandled_command = functions.empty()
+    return true
```
Return values for this function are ignored, safe to remove.

z16 commented on addons/GearSwap/libs/closetCleaner.lua:
```diff
@@ -0,0 +1,367 @@
+                cc.job_used[k] = " "
+            end
+            for i,s in ipairs(ccignore) do
+                if windower.wc_match(gearswap.res.items[k].en, s) or string.match(gearswap.res.items[k].en, s) then
```
At this and other points you use `en` instead of `english` or `name` (same with `english_log` and `name_log` instead of `enl`). And see the comment above about recommending only one matching algorithm. If you go for `name` over `english`, try to do so consistently.

z16 reviewed (changes requested):
Looks like some of your changes didn't make it in or they got lost somewhere along the way. I reviewed it again and mentioned a few things. As before, most of what I said is optional, but I'd like you to adjust the copyright notice and resource usage before I accept this. And I will also highly recommend to remove the default `ccConfig.lua` from the `data` repository, see the specific comment for details.

Fox214 commented:
Ok, I think I missed 1 or two updates from your original note (en vs 
english) not sure how. I have them fixed now.

What should I do with the config file, remove ccConfig.lua and move the 
example one to what directory? (once this is answered i'll recommit, i 
haven't pushed any changes)

I am Brimstone (well it was my original character's name).

The cc.sandbox.windower.register_event= error:apply('Registering events 
disabled for CC') did not work (it crashes on lua with register events 
in them) but removing the () worked and I believe fixed a bug so I think 
that is good.

I'm a unix guy backslash is an escape character and not a path delimiter 
:). Yeah I know the game is in windows, I don't recall if i had issues 
or I just did that to appease my sensibilities (obviously its working 
though).

------ Original Message ------
From: "Stjepan Bakrac" <notifications@github.com>
To: "Windower/Lua" <Lua@noreply.github.com>
Cc: "Fox214" <steven.fox@zoho.com>; "Author" <author@noreply.github.com>
Sent: 11/21/2017 5:51:42 PM
Subject: Re: [Windower/Lua] closetCleaner lib (#1558)

>@z16 requested changes on this pull request.
>
>Looks like some of your changes didn't make it in or they got lost 
>somewhere along the way. I reviewed it again and mentioned a few 
>things. As before, most of what I said is optional, but I'd like you to 
>adjust the copyright notice and resource usage before I accept this. 
>And I will also highly recommend to remove the default ccConfig.lua 
>from the data repository, see the specific comment for details.
>
>--------------------------------------------------------------------------------
>In addons/GearSwap/data/ccConfig.lua 
><https://github.com/Windower/Lua/pull/1558#discussion_r152419654>:
>
> > @@ -0,0 +1,14 @@
>+-- Jobs you want to execute with, recomment put all active jobs you have lua for will look for <job>.lua or <playername>_<job>.lua files
>+-- delete jobs you do not play or have lua files for
>+ccjobs = { 'BLM', 'BLU', 'BRD', 'BST', 'COR', 'DRG', 'DNC', 'DRK', 'GEO', 'MNK', 'NIN', 'PLD', 'PUP', 'RDM', 'RNG', 'RUN', 'SAM', 'SCH', 'SMN', 'THF', 'WAR', 'WHM' }
>+-- Put any items in your inventory here you don't want to show up in the final report
>+-- recommended for furniture, food, meds, pop items or any gear you know you want to keep for some reason
>+-- use * for wildcard matching. see ccConfig_example.lua for examples
>+ccignore = { }
>+-- This is the most use of an item you want to show up in the report
>+-- Set to nil or delete for unlimited
>+ccmaxuse = nil
>+-- List bags you want to not check against, needs to match "Location" column in <player>_report.txt
>+skipBags = S{ 'Temporary' }
>+-- this prints out the _sets _ignored and _inventory files
>+ccDebug = false
>
>This file shouldn't be there. I'd trust people to be competent enough 
>to look at the example config and figure it out, although I would also 
>move that out of the data folder. The problem with the data folder is 
>that anything in there will never be updated. So if you ever adjust or 
>expand the mechanism people's files will not be overwritten (because 
>they're meant to be user created/edited files). This includes the 
>example file.
>
>--------------------------------------------------------------------------------
>In addons/GearSwap/data/ccConfig.lua 
><https://github.com/Windower/Lua/pull/1558#discussion_r152419788>:
>
> > @@ -0,0 +1,14 @@
>+-- Jobs you want to execute with, recomment put all active jobs you have lua for will look for <job>.lua or <playername>_<job>.lua files
>+-- delete jobs you do not play or have lua files for
>+ccjobs = { 'BLM', 'BLU', 'BRD', 'BST', 'COR', 'DRG', 'DNC', 'DRK', 'GEO', 'MNK', 'NIN', 'PLD', 'PUP', 'RDM', 'RNG', 'RUN', 'SAM', 'SCH', 'SMN', 'THF', 'WAR', 'WHM' }
>+-- Put any items in your inventory here you don't want to show up in the final report
>+-- recommended for furniture, food, meds, pop items or any gear you know you want to keep for some reason
>+-- use * for wildcard matching. see ccConfig_example.lua for examples
>+ccignore = { }
>+-- This is the most use of an item you want to show up in the report
>+-- Set to nil or delete for unlimited
>+ccmaxuse = nil
>+-- List bags you want to not check against, needs to match "Location" column in <player>_report.txt
>+skipBags = S{ 'Temporary' }
>
>All the other values in here have cc prefixed (why?), but this one 
>doesn't. Is this an oversight?
>
>--------------------------------------------------------------------------------
>In addons/GearSwap/data/ccConfig_example.lua 
><https://github.com/Windower/Lua/pull/1558#discussion_r152420200>:
>
> > +-- use * for anything.
>+ccignore = { "Rem's Tale*", "Storage Slip *", "Deed of*", "* Virtue", "Dragua's Scale",
>+            "Glittering Yarn", "Dim. Ring*", "Cupboard", "*VCS*", "*Abjuration*", "* Organ",
>+            "Mecisto. Mantle", "Homing Ring", "%a+ Plans", "Orblight", "Yellow 3D Almirah",
>+            "%a+ Statue", "Luminion Chip", ".* Mannequin", "R. Bamboo Grass", "Coiled Yarn",
>+            "Stationery Set", "%a+ Flag", "Bonbori", "Imperial Standard", "%a+ Bed", "Adamant. Statue",
>+            "Festival Dolls", "Taru Tot Toyset", "Bookshelf", "Guild Flyers", "San d'Orian Tree",
>+            ".*Signet Staff", "Capacity Ring", "Facility Ring", "Trizek Ring", "Bam. Grass Basket", "Portafurnace",
>+            "%a+'s Sign", "%a+'s Apron", "Toolbag .*", "%a+ Crepe", "%a+ Sushi.*", ".* Stable Collar", "Plovid Effluvium",
>+            "Gem of the %a+", "Inoshishinofuda", "Ichigohitofuri", "Chonofuda", "Shikanofuda", "Shihei", "Midras.s Helm .1", "Cobra Staff", "Ram Staff", "Fourth Staff",
>+            "Warp Ring", "Chocobo Whistle", "Warp Cudgel", "%a+ Broth", "Tavnazian Ring", "Official Reflector",
>+            "Pashhow Ring", "Dredger Hose", "Trench Tunic", "Sanjaku.Tenugui", "Katana.kazari", "Sanjaku.Tenugui",
>+            "Carver's Torque", "Kabuto.kazari", "Etched Memory", "Trbl. Hatchet", "Maat's Cap", "Trbl. Pickaxe",
>+            "Olduum Ring", "Linkpearl", "Caliber Ring", "Vana'clock", "Signal Pearl", "Windy Greens", "Nexus Cape",
>+            "Wastebasket", "Sickle", "Carpenter's Gloves", "Shinobi.Tabi", "Guardian Board", "%a+ Pie", "Pet Food %a+",
>+            "Reraise %a+", "P. %a+ Card", "Diorite", "Abdhaljs Seal", "Kin's Scale", "*Curry*", "Crepe B. Helene",
>
>Personally I would use one matching mechanism, i.e. windower.wc_match 
>or string.match, not mix both. I'd recommend wildcard matching over Lua 
>patterns when dealing with users.
>
>--------------------------------------------------------------------------------
>In addons/GearSwap/libs/closetCleaner.lua 
><https://github.com/Windower/Lua/pull/1558#discussion_r152420363>:
>
> > @@ -0,0 +1,367 @@
>+--Copyright (c) 2016-2017, Brimstone
>
>Please adjust (c) to ©. Are you Brimstone, or is that a copy/paste 
>oversight?
>
>--------------------------------------------------------------------------------
>In addons/GearSwap/libs/closetCleaner.lua 
><https://github.com/Windower/Lua/pull/1558#discussion_r152420491>:
>
> > +
>+--THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
>+--ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
>+--WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
>+--DISCLAIMED. IN NO EVENT SHALL Brimstone BE LIABLE FOR ANY
>+--DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
>+--(INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
>+--LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
>+--ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
>+--(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
>+--SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
>+
>+-- _addon.name = 'closetCleaner'
>+-- _addon.version = '0.9'
>+-- _addon.author = 'Brimstone'
>+-- _addon.commands = {'cc','closetCleaner'}
>
>Was this originally its own addon, and that's why it's commented out? 
>Any reason to keep it around?
>
>--------------------------------------------------------------------------------
>In addons/GearSwap/libs/closetCleaner.lua 
><https://github.com/Windower/Lua/pull/1558#discussion_r152420969>:
>
> > +--DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
>+--(INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
>+--LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
>+--ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
>+--(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
>+--SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
>+
>+-- _addon.name = 'closetCleaner'
>+-- _addon.version = '0.9'
>+-- _addon.author = 'Brimstone'
>+-- _addon.commands = {'cc','closetCleaner'}
>+
>+local cc = {}
>+cc.sandbox = {}
>+cc.sandbox.windower = setmetatable({}, {__index = windower})
>+cc.sandbox.windower.register_event = functions.empty()
>
>Are you trying to disable register_event? If so, you'd want 
>functions.empty, not functions.empty(). The latter is the same as using 
>nil.
>
>Any particular reason to do that? I would probably throw an error 
>instead, so people actually know not to use it, instead of nothing 
>happening:
>
>cc.sandbox.windower.register_event= error:apply('Registering events disabled for CC')
>--------------------------------------------------------------------------------
>In addons/GearSwap/libs/closetCleaner.lua 
><https://github.com/Windower/Lua/pull/1558#discussion_r152422701>:
>
> > +
>+-- _addon.name = 'closetCleaner'
>+-- _addon.version = '0.9'
>+-- _addon.author = 'Brimstone'
>+-- _addon.commands = {'cc','closetCleaner'}
>+
>+local cc = {}
>+cc.sandbox = {}
>+cc.sandbox.windower = setmetatable({}, {__index = windower})
>+cc.sandbox.windower.register_event = functions.empty()
>+cc.sandbox.windower.raw_register_event = functions.empty()
>+cc.sandbox.windower.register_unhandled_command = functions.empty()
>+
>+register_unhandled_command(function(...)
>+    local cmds = {...}
>+    for _,v in ipairs(cmds) do
>
>Are you sure this is what you want? This would trigger even if cc was a 
>third parameter in an unhandled command that was meant to be handled by 
>something else.
>
>I'd change the function signature and setup to this:
>
>register_unhandled_command(function(command)
>     command = command and command:lower() ornilif command ~='cc and command ~= 'closetcleaner' then        return    end    --[[ rest of the function ]]end)
>--------------------------------------------------------------------------------
>In addons/GearSwap/libs/closetCleaner.lua 
><https://github.com/Windower/Lua/pull/1558#discussion_r152422876>:
>
> > +register_unhandled_command(function(...)
>+    local cmds = {...}
>+    for _,v in ipairs(cmds) do
>+        if S{'closetcleaner','cc'}:contains(v:lower()) then
>+            setmetatable(cc.sandbox, {__index = gearswap.user_env})
>+            cc.sandbox.itemsBylongName = T{}
>+            cc.sandbox.itemsByName = T{}
>+            cc.sandbox.inventoryGear = T{}
>+            cc.sandbox.gsGear = T{}
>+            for k,v in pairs(gearswap.res.items) do
>+                cc.sandbox.itemsBylongName[gearswap.res.items[k].name_log:lower()] = k
>+                cc.sandbox.itemsByName[gearswap.res.items[k].name:lower()] = k
>+            end
>+            cc.sandbox.jobs = {}
>+            for k,v in pairs(gearswap.res.jobs) do
>+                cc.sandbox.jobs[gearswap.res.jobs[k].ens] = k
>
>Please use english_short instead of en.
>
>--------------------------------------------------------------------------------
>In addons/GearSwap/libs/closetCleaner.lua 
><https://github.com/Windower/Lua/pull/1558#discussion_r152423016>:
>
> > +            cc.sandbox.itemsBylongName = T{}
>+            cc.sandbox.itemsByName = T{}
>+            cc.sandbox.inventoryGear = T{}
>+            cc.sandbox.gsGear = T{}
>+            for k,v in pairs(gearswap.res.items) do
>+                cc.sandbox.itemsBylongName[gearswap.res.items[k].name_log:lower()] = k
>+                cc.sandbox.itemsByName[gearswap.res.items[k].name:lower()] = k
>+            end
>+            cc.sandbox.jobs = {}
>+            for k,v in pairs(gearswap.res.jobs) do
>+                cc.sandbox.jobs[gearswap.res.jobs[k].ens] = k
>+            end
>+            if not windower.dir_exists(windower.addon_path..'report') then
>+                windower.create_dir(windower.addon_path..'report')
>+            end
>+            local path = windower.addon_path:gsub('\\','/')
>
>Our of curiosity, did you have issues using \ over /?
>
>--------------------------------------------------------------------------------
>In addons/GearSwap/libs/closetCleaner.lua 
><https://github.com/Windower/Lua/pull/1558#discussion_r152423093>:
>
> > +            end
>+            if not windower.dir_exists(windower.addon_path..'report') then
>+                windower.create_dir(windower.addon_path..'report')
>+            end
>+            local path = windower.addon_path:gsub('\\','/')
>+            path = path..'report/'..player.name
>+            require 'ccConfig'
>+            cc.run_report(path)
>+            cc.sandbox = nil
>+        end
>+    end
>+    cc.sandbox = {}
>+    cc.sandbox.windower = setmetatable({}, {__index = windower})
>+    cc.sandbox.windower.register_event = functions.empty()
>+    cc.sandbox.windower.raw_register_event = functions.empty()
>+    cc.sandbox.windower.register_unhandled_command = functions.empty()
>
>See above about function.empty.
>
>--------------------------------------------------------------------------------
>In addons/GearSwap/libs/closetCleaner.lua 
><https://github.com/Windower/Lua/pull/1558#discussion_r152423180>:
>
> > +            if not windower.dir_exists(windower.addon_path..'report') then
>+                windower.create_dir(windower.addon_path..'report')
>+            end
>+            local path = windower.addon_path:gsub('\\','/')
>+            path = path..'report/'..player.name
>+            require 'ccConfig'
>+            cc.run_report(path)
>+            cc.sandbox = nil
>+        end
>+    end
>+    cc.sandbox = {}
>+    cc.sandbox.windower = setmetatable({}, {__index = windower})
>+    cc.sandbox.windower.register_event = functions.empty()
>+    cc.sandbox.windower.raw_register_event = functions.empty()
>+    cc.sandbox.windower.register_unhandled_command = functions.empty()
>+    return true
>
>Return values for this function are ignored, safe to remove.
>
>--------------------------------------------------------------------------------
>In addons/GearSwap/libs/closetCleaner.lua 
><https://github.com/Windower/Lua/pull/1558#discussion_r152423458>:
>
> > +    if ccDebug then
>+        ignoredReportName = path..'_ignored.txt'
>+        f2 = io.open(ignoredReportName,'w+')
>+        f2:write('closetCleaner ignored Report:\n')
>+        f2:write('=====================\n\n')
>+        cc.print_row(f2, data, form)
>+        cc.print_break(f2, form)
>+    end
>+    for k,v in cc.spairs(cc.sandbox.gsGear, function(t,a,b) return t[b] > t[a] end) do
>+        if ccmaxuse == nil or v <= ccmaxuse then
>+            printthis = 1
>+            if not cc.job_used[k] then
>+                cc.job_used[k] = " "
>+            end
>+            for i,s in ipairs(ccignore) do
>+                if windower.wc_match(gearswap.res.items[k].en, s) or string.match(gearswap.res.items[k].en, s) then
>
>At this and other points you use en instead of english or name (same 
>with english_log and name_log instead of enl). And see the comment 
>above about recommending only one matching algorithm. If you go for 
>name over english, try to do so consistently.
>
>—
>You are receiving this because you authored the thread.
>Reply to this email directly, view it on GitHub 
><https://github.com/Windower/Lua/pull/1558#pullrequestreview-78280191>, 
>or mute the thread 
><https://github.com/notifications/unsubscribe-auth/ARqYDxl-lnulunGQf5YHKmfYzyaIQqoJks5s41P-gaJpZM4O8mdc>.
>

z16 commented:
Yeah, both `\` and `/` work, was just curious.

I would recommend using the `config` library for configurations in the data folder. It will automatically create the file from default settings you pass into it, if it's missing, and can even amend single settings if you add new ones in the future.

Fox214 commented:
Ok 1 question about the config library, since its in gearswap/data 
folder and gets named settings.xml what happens if gearswap or someone 
else creates a similar library for gearswap that uses the config library 
will they conflict?

------ Original Message ------
From: "Stjepan Bakrac" <notifications@github.com>
To: "Windower/Lua" <Lua@noreply.github.com>
Cc: "Fox214" <steven.fox@zoho.com>; "Author" <author@noreply.github.com>
Sent: 12/2/2017 5:05:03 AM
Subject: Re: [Windower/Lua] closetCleaner lib (#1558)

>Yeah, both \ and / work, was just curious.
>
>I would recommend using the config library for configurations in the 
>data folder. It will automatically create the file from default 
>settings you pass into it, if it's missing, and can even amend single 
>settings if you add new ones in the future.
>
>—
>You are receiving this because you authored the thread.
>Reply to this email directly, view it on GitHub 
><https://github.com/Windower/Lua/pull/1558#issuecomment-348681820>, or 
>mute the thread 
><https://github.com/notifications/unsubscribe-auth/ARqYD7aWtOuRTvjicWm8P8tNFOx8aZZxks5s8SDPgaJpZM4O8mdc>.
>

z16 commented:
You can specify a filename:
```lua
settings = config.load('ccConfig.xml', defaults) -- or whatever
```

Otherwise yes, it would conflict.

Fox214 commented:
Well I updated to the xml file, but i really hate the way the ccignore 
list works, because this is likely to be a long list for a lot of folks.

<ccignore>
     <1>someth*ng</1>
     <2>foo*bar</2>
</ccignore>

Is a lot more verbose to edit than:
  ccignore = {"someth*ng", "foo*bar"}

(7:2 character ratio not to mention matching/non-repeat issue, I 
wouldn't want to type out 50 items that way)

I tried using it as S{} table that fixes the Xml but just doesn't work 
in the lua then.

Any ideas?

------ Original Message ------
From: "Stjepan Bakrac" <notifications@github.com>
To: "Windower/Lua" <Lua@noreply.github.com>
Cc: "Fox214" <steven.fox@zoho.com>; "Author" <author@noreply.github.com>
Sent: 12/5/2017 12:37:11 AM
Subject: Re: [Windower/Lua] closetCleaner lib (#1558)

>You can specify a filename:
>
>settings = config.load('ccConfig.xml', defaults) -- or whatever
>Otherwise yes, it would conflict.
>
>—
>You are receiving this because you authored the thread.
>Reply to this email directly, view it on GitHub 
><https://github.com/Windower/Lua/pull/1558#issuecomment-349201025>, or 
>mute the thread 
><https://github.com/notifications/unsubscribe-auth/ARqYD63soIGOG52eVL9t_WDrLkXzEz_5ks5s9NaHgaJpZM4O8mdc>.
>

z16 commented:
It should work fine with `S{}` tables (and `L{}`-tables for that matter), other addons use that too (like Treasury) and it results in a normal, comma-separated list of strings. I would try to figure out why that doesn't work for you. If you want you [gist](https://gist.github.com/) the code you have that isn't working for you, then I can have a look.

Commit:
Updates done

using xml settings now, wrote out names and some other minor fixes.

Fox214 commented:
I was able to figure it out new commit here: 
https://github.com/Windower/Lua/pull/1558/commits/20fc7e50d6a42a21bb4fa3441ab714e9cd3e0d91

(had to change the iterator with S{}) I think its ok now.

------ Original Message ------
From: "Stjepan Bakrac" <notifications@github.com>
To: "Windower/Lua" <Lua@noreply.github.com>
Cc: "Fox214" <steven.fox@zoho.com>; "Author" <author@noreply.github.com>
Sent: 12/6/2017 12:15:43 AM
Subject: Re: [Windower/Lua] closetCleaner lib (#1558)

>It should work fine with S{} tables (and L{}-tables for that matter), 
>other addons use that too (like Treasury) and it results in a normal, 
>comma-separated list of strings. I would try to figure out why that 
>doesn't work for you. If you want you gist <https://gist.github.com/> 
>the code you have that isn't working for you, then I can have a look.
>
>—
>You are receiving this because you authored the thread.
>Reply to this email directly, view it on GitHub 
><https://github.com/Windower/Lua/pull/1558#issuecomment-349536507>, or 
>mute the thread 
><https://github.com/notifications/unsubscribe-auth/ARqYD4N9uiEDC6wdMomRdDFhEe-QemUMks5s9iL_gaJpZM4O8mdc>.
>

Fox214 commented:
Awesome just out of curiosity any idea when it might be live?

------ Original Message ------
From: "Stjepan Bakrac" <notifications@github.com>
To: "Windower/Lua" <Lua@noreply.github.com>
Cc: "Fox214" <steven.fox@zoho.com>; "Author" <author@noreply.github.com>
Sent: 12/9/2017 2:25:34 AM
Subject: Re: [Windower/Lua] closetCleaner lib (#1558)

>Merged #1558 <https://github.com/Windower/Lua/pull/1558>.
>
>—
>You are receiving this because you authored the thread.
>Reply to this email directly, view it on GitHub 
><https://github.com/Windower/Lua/pull/1558#event-1379931516>, or mute 
>the thread 
><https://github.com/notifications/unsubscribe-auth/ARqYD-32-npcrQGYOp2eagcPOmswkukYks5s-jXugaJpZM4O8mdc>.
>

z16 commented:
Now :)
